Source/Engine/PackageBuilder/PatternLinker.cs

```diff
@@ -22,57 +25,100 @@ internal struct PatternReferenceInContext
             public PatternSyntax PatternContext;
         }
 
+        private static readonly HashSet<string> StandardPatternNames;
+        private readonly Func<string, PackageSyntax> fPackageProvider;
+        private readonly ILinkerCache fLinkerCache;
+        private readonly Stack<string> fDependencyStack;
+        private string fBaseDirectory;
         private Dictionary<string, PatternSyntax> fPatternByName;
+        private Dictionary<string, RequiredPackageSyntax> fRequiredPackageByFilePath;
+        private Dictionary<string, RequiredPackageSyntax> fRequiredPackageByPatternName;
+        private Dictionary<RequiredPackageSyntax, Dictionary<string, List<string>>> fDuplicatePatternsByRequiredPackage;
         private List<PatternReferenceInContext> fPatternReferences;
         private PatternSyntax fCurrentPattern;
-        private readonly string fBaseDirectory;
-        private readonly IPackageLoader fRequiredPackageLoader;
+        private List<Error> fErrors;
 
-        public PatternLinker()
-            : this(null, null)
+        static PatternLinker()
+        {
+            StandardPatternNames = new HashSet<string>(Syntax.StandardPattern.StandardPatterns.Select(x => x.FullName));
+        }
+
+        public PatternLinker(Func<string, PackageSyntax> packageProvider)
+            : this(packageProvider, linkerCache: null)
         {
         }
         
-        public PatternLinker(string baseDirectory, IPackageLoader requiredPackageLoader)
+        public PatternLinker(Func<string, PackageSyntax> packageProvider, ILinkerCache linkerCache)
         {
-            fBaseDirectory = baseDirectory;
-            fRequiredPackageLoader = requiredPackageLoader;
+            fPackageProvider = packageProvider ?? throw new ArgumentNullException(nameof(packageProvider));
```

Comment:
Please use same null check pattern as in other places:
```csharp
if (value == null)
    throw new ArgumentNullException(nameof(value));
this.value = value;
```

Reply:
Allow `null` here and throw exception on usage. It simplifies creating `PatternLinker` in tests: no need to provide delegate which always throws exception.